File: src/main/java/com/example/reactive/chapter2/observer/Subject.java
package com.example.reactive.chapter2.observer;

import java.util.ArrayList;
import java.util.List;

public class Subject implements Publisher<Event> {

    private List<Observer> observers = new ArrayList<>();

    @Override
    public void registerObserver(Observer observer) {

        observers.add(observer);
    }

    @Override
    public void unregisterObserver(Observer observer) {

        observers.remove(observer);
    }

    @Override
    public void notifyObservers(Event event) {

        observers.forEach(observer -> observer.observe(event));
    }
}

File: src/main/kotlin/wmp/edu/reactive/observer/Subject.kt
package wmp.edu.reactive.observer

class Subject {
    private val observable = mutableListOf<Observer>()

    fun notify(event: Event) {
        for (observer in observable) {
            observer.notify(event)
        }
    }

    fun registerObserver(observer: Observer) {
        observable.add(observer)
    }

    fun removeObserver(observer: Observer) {
        observable.remove(observer)
    }
}
